### PR TITLE
🐛 Handle Wise SCA requests

### LINKED ISSUE #2144: Wise API requires strong customer authentication
**Describe the bug**
When using the wise node (and the wise trigger node respectively) the requests will fail after some time.
Wise requires solving a challenge by signing a string found in the http headers with the clients private RSA key.

See https://api-docs.wise.com/#webhook-events-transfer-status-change-event

**To Reproduce**
Use a wise node with an european wise account.

**Expected behavior**
Wise node should solve the RSA challenge.

**Environment (please complete the following information):**
 - n8n Version 0.135.3

### PR DESCRIPTION
**Pull request by pemontto**
Fixes #2144

This PR adds [Strong Customer Authentication (SCA)](https://api-docs.transferwise.com/#strong-customer-authentication) to the Wise node and extra context to recipients that could previously be hard to distinguish.

SCA is only required on getStatements, and Funds transfer request. It is disabled when using test endpoints which makes it difficult to test.

To support SCA we need to:

- get back a 403 response
- extract a one-time token from the headers `X-2FA-Approval`
- create a SHA256 signature with an RSA key
- resend the request with new headers `X-2FA-Approval` and `X-Signature`

### 🚧 403/error responses
By default all of the request helpers treat non-success codes as errors. To get back the error response with headers I needed to add the Axios `validateStatus` option to the `IHttpRequestOptions`. In developing nodes I've sometimes wondered if I could get all responses and decide what constitutes a failure condition. Is this something I've missed that is already possible?

### 🔑 Handle one-time tokens
To enable signing a token I've added a `privateKey` to the Wise credentials. The basic instructions for configuring this are

1. Generate an RSA key pair 
  ```shell
  openssl genrsa -out wise-n8n.pem 2048
  openssl rsa -pubout -in wise-n8n.pem -out wise-n8n-public.pem
  ```
2. In the `API token` section of your [profile](https://wise.com/settings/) go to `Manage your public keys`
3. Upload the `wise-n8n-public.pem` public key
4. Add the contents of `wise-n8n.pem` to the `Private Key (Optional)` credential field.

### 👩‍💻 Recipient display
This PR also updates the recipient loadOptions to filter for active accounts, and show more context around the account, like country and currency.

Before (first 6 options are the same name):
n8n_io_-_Workflow_Automation

After (shows currency, country, and account type):
n8n_-_▶️_My_workflow_4

### CODE REVIEW

**Comment by pemontto** [comment]
This probably doesn't strictly require a new `wiseSCAApiRequest` function, we can probably just check for SCA failure regardless of the request and cut down on duplicated code. Happy to take feedback 👍

File: packages/core/src/NodeExecuteFunctions.ts [hunk 1]
```diff
@@ -729,6 +729,10 @@ function convertN8nRequestToAxios(n8nRequest: IHttpRequestOptions): AxiosRequest
 		axiosRequest.headers['User-Agent'] = 'n8n';
 	}
 
+	if (n8nRequest.validateStatus !== undefined) {
+		axiosRequest.validateStatus = n8nRequest.validateStatus;
+	}
```
Comment (ivov) [inline packages/core/src/NodeExecuteFunctions.ts:734]: > By default all of the request helpers treat non-success codes as errors. To get back the error response with headers I needed to add the Axios `validateStatus` option to the `IHttpRequestOptions`. In developing nodes I've sometimes wondered if I could get all responses and decide what constitutes a failure condition. Is this something I've missed that is already possible?

@krynble If you can confirm that you meant for this to be done with `validateStatus`.

File: packages/nodes-base/credentials/WiseApi.credentials.ts [hunk 2]
```diff
@@ -30,5 +30,15 @@ export class WiseApi implements ICredentialType {
 				},
 			],
 		},
+		{
+			displayName: 'Private Key (Optional)',
+			name: 'privateKey',
+			type: 'string',
+			default: '',
+			description: 'Optional private key used for Strong Customer Authentication (SCA)',
```
Comment (ivov) [inline packages/nodes-base/credentials/WiseApi.credentials.ts:38]: Would be useful to mention here the two SCA operations.
```suggestion
			description: 'Optional private key used for Strong Customer Authentication (SCA). Only needed for X and Y.',
```

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts [hunk 3]
```diff
@@ -65,6 +70,95 @@ export async function wiseApiRequest(
 	}
 }
 
+/**
+ * Make an authenticated SCA API request to Wise.
+ */
+ export async function wiseSCAApiRequest(
```
Comment (ivov) [inline packages/nodes-base/nodes/Wise/GenericFunctions.ts:76]: > This probably doesn't strictly require a new `wiseSCAApiRequest` function, we can probably just check for SCA failure regardless of the request and cut down on duplicated code.

Agree

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts [hunk 4]
```diff
@@ -65,6 +70,95 @@ export async function wiseApiRequest(
...
+			'Authorization': `Bearer ${apiToken}`,
+		},
+		method,
+		url: `${rootUrl}${endpoint}`,
+		qs,
+		body,
+		json: true,
+		returnFullResponse: true,
+		validateStatus: (status) => {
+			return (status >= 200 && status < 300) || status === 403; // default
```
Comment (ivov) [inline packages/nodes-base/nodes/Wise/GenericFunctions.ts:106]: What does `default` mean here?

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts [hunk 5]
```diff
@@ -65,6 +70,95 @@ export async function wiseApiRequest(
...
+			'x-2fa-approval': oneTimeToken,
+		};
+		try {
+			response = await this.helpers.httpRequest!(options);
+			return response.body;
+		} catch (error) {
+			throw new NodeApiError(this.getNode(), error);
+		}
+	} else {
+		throw new NodeApiError(this.getNode(), {...response.body},);
```
Comment (ivov) [inline packages/nodes-base/nodes/Wise/GenericFunctions.ts:158]: Why make a copy of the body? And why only the body? If this else branch is reached, it looks like the only possibility is that it was a 403 lacking the expected SCA header, so it'd be useful for the error to contain the received headers.

File: packages/nodes-base/nodes/Wise/Wise.node.ts [hunk 6]
```diff
@@ -143,10 +144,11 @@ export class Wise implements INodeType {
...
 
-				return recipients.map(({ id, accountHolderName }: Recipient) => ({
-					name: accountHolderName,
-					value: id,
-				}));
+				return recipients.filter(({active} : Recipient) => active)
+					.map(({ id, accountHolderName, currency, country, type }: Recipient) => ({
+						name: `[${currency}] ${accountHolderName} - (${country !== null ? country + ' - ' : '' }${type})`,
+						value: id,
+					}));
```
Comment (ivov) [inline packages/nodes-base/nodes/Wise/Wise.node.ts:151]: `reduce` can do this in a single pass.

Comment (ivov) [reply packages/nodes-base/nodes/Wise/Wise.node.ts:151]: `as Recipient[]` at L145 so the item type is inferrable in the loop.

Comment (ivov) [reply packages/nodes-base/nodes/Wise/Wise.node.ts:151]: `country !== null` will always be true according to the typing. Either the typing or the check needs adjusting.

https://github.com/n8n-io/n8n/blob/189a6e61e64921f5e193c7f0e43a1b8fa57c593e/packages/nodes-base/nodes/Wise/GenericFunctions.ts#L209-L216

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts [hunk 7]
```diff
@@ -65,6 +70,95 @@ export async function wiseApiRequest(
...
+	method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'HEAD',
+	endpoint: string,
+	body: IDataObject = {},
+	qs: IDataObject = {},
+	option: IDataObject = {},
+) {
+	const { apiToken, environment, privateKey } = await this.getCredentials('wiseApi') as {
+		apiToken: string,
+		environment: 'live' | 'test',
+		privateKey: string | null,
```
Comment (ivov) [inline packages/nodes-base/nodes/Wise/GenericFunctions.ts:87]:
```suggestion
		privateKey?: string,
```

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts [hunk 8]
```diff
@@ -65,6 +70,95 @@ export async function wiseApiRequest(
...
+
+	let response;
+	try {
+		response = await this.helpers.httpRequest!(options);
+	} catch (error) {
+		throw new NodeApiError(this.getNode(), error);
+	}
+
+	if (response.statusCode === 200) {
+		return response.body;
```
Comment (ivov) [inline packages/nodes-base/nodes/Wise/GenericFunctions.ts:130]: Note to myself to check when testing. The original calls `this.helpers.request` and returns the response unaltered. This version calls `this.helpers.httpRequest` and returns `response.body`. Same at L153. Does this change preserve the shape of the original response?

**Review by ivov — COMMENTED** [review]
Thank you for taking this on.

File: packages/nodes-base/credentials/WiseApi.credentials.ts (hunk 2 above)
Comment (pemontto) [reply packages/nodes-base/credentials/WiseApi.credentials.ts]: Updated to `Only required to retrieve statements, and execute transfers`

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts (hunk 4 above)
Comment (pemontto) [reply packages/nodes-base/nodes/Wise/GenericFunctions.ts]: Bad copy/paste, sorry!

File: packages/nodes-base/nodes/Wise/Wise.node.ts (hunk 6 above)
Comment (pemontto) [reply packages/nodes-base/nodes/Wise/Wise.node.ts]: Thanks, always have trouble conceptualising reduce. Let me know if the updated change is what you expected.

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts (hunk 7 above)
Comment (pemontto) [reply packages/nodes-base/nodes/Wise/GenericFunctions.ts]: Fixed, I'm not string with typing!

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts (hunk 8 above)
Comment (pemontto) [reply packages/nodes-base/nodes/Wise/GenericFunctions.ts:76]: As above, the `request` helper just returns `response.body`

File: packages/nodes-base/nodes/Wise/GenericFunctions.ts (hunk 3 above)
Comment (pemontto) [reply packages/nodes-base/nodes/Wise/GenericFunctions.ts]: Merged the functions. As far as I can tell the request helper just returns `response.body`, so using the `httpRequest` helper and returning `response.body` is equivalent.

File: packages/nodes-base/nodes/Wise/Wise.node.ts (hunk 6 above)
Comment (ivov) [reply packages/nodes-base/nodes/Wise/Wise.node.ts]: Exactly. 

Minor optional suggestions:
- The type of the previous value should be provided [as a generic](https://github.com/microsoft/TypeScript/blob/afee8bf9682dfb3ed927fdc7756d96a19851b811/lib/lib.es5.d.ts#L1386 ) to `reduce`
- The type of the current value can be omitted as it is inferrable from the array
- In `Recipient`, `currency` should be `string | null`  for the check to make sense

```ts
return recipients.reduce<INodePropertyOptions[]>((activeRecipients, {
	active,
	id,
	accountHolderName,
	currency,
	country,
	type,
}) => {
	if (active) {
		const recipient = {
			name: `[${currency}] ${accountHolderName} - (${country !== null ? country + ' - ' : '' }${type})`,
			value: id,
		};
		activeRecipients.push(recipient);
	}
	return activeRecipients;
}, []);
```

File: packages/nodes-base/nodes/ItemLists/ItemLists.node.ts [hunk 9]
```diff
@@ -95,7 +95,7 @@ export class ItemLists implements INodeType {
 				name: 'fieldToSplitOut',
 				type: 'string',
 				default: '',
-				required: true,
+				required: false,
```
[the remark this answers is not in the document]
Comment (pemontto) [reply packages/nodes-base/nodes/ItemLists/ItemLists.node.ts]: Oops, fixed that one.

File: packages/nodes-base/nodes/Wise/Wise.node.ts (hunk 6 above)
Comment (ivov) [reply packages/nodes-base/nodes/Wise/Wise.node.ts]: Exactly.

Minor optional suggestions:
- The type for the accumulator is provided [as a generic](https://github.com/microsoft/TypeScript/blob/afee8bf9682dfb3ed927fdc7756d96a19851b811/lib/lib.es5.d.ts#L1386) to `reduce`
- The type for the current value can be omitted as it is inferrable from the array
- In `Recipient`, `country` should be `string | null` for the check to make sense.

```ts
return recipients.reduce<INodePropertyOptions[]>((activeRecipients, {
	active,
	id,
	accountHolderName,
	currency,
	country,
	type,
}) => {
	if (active) {
		const recipient = {
			name: `[${currency}] ${accountHolderName} - (${country !== null ? country + ' - ' : '' }${type})`,
			value: id,
		};
		activeRecipients.push(recipient);
	}
	return activeRecipients;
}, []);
```

**Comment by ivov** [comment]
Looks good to me. Waiting for @krynble to confirm.

File: packages/nodes-base/nodes/Wise/Wise.node.ts (hunk 6 above)
Comment (pemontto) [reply packages/nodes-base/nodes/Wise/Wise.node.ts]: Amazing, I have some reading to do on generics. Have tested and updated the commit with your suggestions.

File: packages/core/src/NodeExecuteFunctions.ts (hunk 1 above)
Comment (krynble) [reply packages/core/src/NodeExecuteFunctions.ts:734]: Hey @ivov as you mentioned, the new `httpRequest` helper does not provide a way to return the response data when a request is not 2xx.

I believe the best way is to provide a new option named `ignoreHttpStatusErrors` or another name with a behavior similar to the existing `returnFullResponse` option that can be see in the `httpRequest` function in this same file.

Comment (pemontto) [reply packages/core/src/NodeExecuteFunctions.ts:734]: @ivov, @krynble here's a version using `ignoreHttpStatusErrors` instead:
https://github.com/n8n-io/n8n/compare/master...pemontto:wise-sca2?expand=1

Main change is 
```js
if (n8nRequest.ignoreHttpStatusErrors) {
    axiosRequest.validateStatus = () => true;
}
```

**Comment by krynble** [comment]
Hey @pemontto and @ivov 

I checked the other branch named `wise-sca2` in Pemontto's account and it looks good.

You can create a now pr and close this one in favor of it.

**Comment by pemontto** [comment]
@krynble I've pushed that branch commit onto this branch to keep it with the PR conversation.

**Comment by janober** [comment]
Thanks a lot for your contribution. Got merged.

**Comment by janober** [comment]
Got released with `n8n@0.166.0`

**Comment by Vad1mo** [comment]
The documentation in this PR should be present in https://docs.n8n.io/integrations/builtin/credentials/wise/

**Comment by Vad1mo** [comment]
@pemontto done -> https://github.com/n8n-io/n8n-docs/pull/2154

**Pull request merged**